### DISCUSSION TITLE
Best Practice for Page Object creation in сross-platform mobile testing: android().ios() vs @AndroidFindBy/@iOSFindBy

### DISCUSSION THREAD

**Discussion by OleksandrOwl** [selenide/selenide#2817] [Q&A] [answered]
Hello,

I’m working on a cross-platform mobile automation project using Appium and am at a crossroads with how to effectively create page objects. The decision is between using:

  1. Platform-specific annotations such as **@AndroidFindBy** for Android and **@iOSFindBy** for iOS within the same page object to handle platform-specific elements.
  2. A combined approach with a custom **android().ios()** method to encapsulate both platforms’ locators in one place.

I would greatly appreciate it if the community could share their experiences and insights on these methods:

  1. Which approach do you find more maintainable and scalable, especially for a project that targets multiple platforms?
  2. Could you provide concrete examples of complex page objects (including widgets and the like) structured using your preferred approach?

Furthermore, I noticed that @asolntsev has recommendations against using PageFactory. I am eager to learn more about the reasoning behind this perspective.

What are the concerns associated with the PageFactory pattern?
What alternatives do you suggest, and what are their pros and cons, in practice?

Thank you for your valuable insights and looking forward to your responses.

**Comment by asolntsev** [comment, marked as answer]
@OleksandrOwl You can use both ways, there is no big difference. 

My objections against page objects with `@FindBy` fields are described here: https://selenide.org/documentation/page-objects.html#what-disadvantages

But it depends on your context.

You can find examples of both style page objects in Selenide own tests: https://github.com/selenide/selenide/tree/main/modules/appium/src/test/java/it/mobile
